fero/analysis.py

```diff
@@ -148,3 +328,314 @@ def make_prediction(
+            )
+            return goal_data["dtype"]
+        except StopIteration:
+            return None
+
+    def _verify_standard_goal(self, goal: dict):
+        """Verifies the goal config relative to the analysis"""
+        goal_name = goal["factor"]["name"]
+
+        # The goal label must be a target or factor
+        if goal_name not in self.target_names + self.factor_names:
+            raise FeroError(f'"{goal_name}" is not a valid goal')
+
+        # If this is a factor makes sure it's not a float
+        if (
+            goal_name not in self.target_names
+            and self._get_factor_dtype(goal_name) != "float"
+        ):
+            raise FeroError("Goal must be a float")
+
+    def _verify_cost_goal(self, goal: dict):
+        """Verify that a cost goal is correct"""
+        for factor in goal["cost_function"]:
+            factor_name = factor["name"]
+            factor_type = self._get_factor_dtype(factor_name)
+            if factor_type is None:
+                raise FeroError(f'"{factor_name}" is not a factor')
+            # Implicitly find missing factors
+            if factor_type != "float":
+                raise FeroError("Cost functions factors must be factors")
+
+    def _verify_constraints(self, constraints: List[dict]):
+        """verify provided constraints are in the analysis"""
+        all_names = self.factor_names + self.target_names
+        for constraint in constraints:
+            constraint_name = constraint["name"]
+            if constraint_name not in all_names:
+                raise FeroError(
+                    f'Constraint "{constraint_name}" is not in this analysis'
+                )
+
+    def _cross_verify_optimization(self, goal: dict, constraints: List[dict]):
+        """Verify the config has the correct combined targets, costs and factors"""
+        is_cost = goal.get("type", None) == "COST"
+
+        cost_factors = []
+        goal_factor = []
+        target_factor = []
+        constraint_factors = []
+        constraint_targets = []
+
+        if is_cost:
+            for factor in goal["cost_function"]:
+                cost_factors.append(factor["name"])
+        else:
+            if goal["factor"]["name"] in self.factor_names:
+                goal_factor.append(goal["goal"])
+            else:
+                target_factor.append(goal["goal"])
+
+        for constraint in constraints:
+            if constraint["name"] in self.factor_names:
+                constraint_factors.append(constraint["name"])
+            else:
+                constraint_targets.append(constraint["name"])
+
+        if len(target_factor + constraint_targets) < 1:
+            raise FeroError("No Targets specified")
+
+        if len(constraint_factors + cost_factors + goal_factor) > 3:
+            raise FeroError(
+                "A maximum of three factors can be specified in an optimization"
+            )
```

Comment:
I _think_ this would not catch the case where the target is set as a goal but no constraints are provided. Should probably check that there is at least 1 constraint.

Reply:
Yep, check added for the 0 constraint case.